In [1]:
import numpy as np
import my_submission as ms
import matplotlib.pyplot as plt

/Users/Ding/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


In [2]:
#from tensorflow.python.client import device_lib
#print(device_lib.list_local_devices())

In [3]:
#ms.save_mnist_dataset()
# Loads the already preprocessed dataset.
#(x_train, y_train), (x_test, y_test) = ms.load_mnist_dataset()
#print('x_train : ', x_train.shape, x_train.dtype)
#print('y_train : ', y_train.shape, y_train.dtype)
#print('x_test : ', x_test.shape, x_test.dtype)
#print('y_test : ', y_test.shape, y_test.dtype)

#x_train, y_train, x_test, y_test = ms.preprocess_mnist_dataset(x_train, y_train, x_test, y_test)
# Saves the preprocessed dataset locally as a .npz file.
#np.savez('mnist_dataset.npz', x_train = x_train, y_train = y_train, x_test = x_test, y_test = y_test)


In [4]:
# Verifying that the split has been done correctly
#print("Unique numbers in train set (shouldn't be any of the 0,1,8 or 9):")
#unique, counts = np.unique(y_train, return_counts=True)
#dict(zip(unique, counts))

In [5]:
# Verifying that the split has been done correctly
#print("Unique numbers in test set:")
#unique, counts = np.unique(y_test, return_counts=True)
#dict(zip(unique, counts))

In [6]:
#ms.build_CNN(x_train, y_train, x_test, y_test)
ms.siamese_network()

input_trainset:  (33149, 28, 28, 1)
target_trainset:  (33149,)
input_testset1:  (8288, 28, 28, 1)
target_testset1:  (8288,)
input_testset2:  (28563, 28, 28, 1)
target_testset2:  (28563,)
input_testset3:  (36851, 28, 28, 1)
target_testset3:  (36851,)
--------------------
training_pairs:  (60792, 2, 28, 28, 1)
training_target:  (60792,)
--------------------
test_pairs_set1:  (14940, 2, 28, 28, 1)
test_target_set1:  (14940,)
--------------------
test_pairs_set2:  (54592, 2, 28, 28, 1)
test_target_set2:  (54592,)
--------------------
test_pairs_set3:  (24900, 2, 28, 28, 1)
test_target_set3:  (24900,)
--------------------
Train on 60792 samples, validate on 14940 samples
Epoch 1/30
60792/60792 [==============================] - 77s 1ms/step - loss: 0.4699 - accuracy: 0.8709 - val_loss: 0.3246 - val_accuracy: 0.9457
Epoch 2/30
60792/60792 [==============================] - 76s 1ms/step - loss: 0.3249 - accuracy: 0.9497 - val_loss: 0.3058 - val_accuracy: 0.9588
Epoch 3/30
60792/60792 [=======

<h1>Results:</h1>
<p>
input_trainset:  (33149, 28, 28, 1)
target_trainset:  (33149,)
input_testset1:  (8288, 28, 28, 1)
target_testset1:  (8288,)
input_testset2:  (28563, 28, 28, 1)
target_testset2:  (28563,)
input_testset3:  (36851, 28, 28, 1)
target_testset3:  (36851,)
--------------------
training_pairs:  (60792, 2, 28, 28, 1)
training_target:  (60792,)
--------------------
test_pairs_set1:  (14940, 2, 28, 28, 1)
test_target_set1:  (14940,)
--------------------
test_pairs_set2:  (54592, 2, 28, 28, 1)
test_target_set2:  (54592,)
--------------------
test_pairs_set3:  (24900, 2, 28, 28, 1)
test_target_set3:  (24900,)
--------------------
Train on 60792 samples, validate on 14940 samples
Epoch 1/30
60792/60792 [==============================] - 77s 1ms/step - loss: 0.4699 - accuracy: 0.8709 - val_loss: 0.3246 - val_accuracy: 0.9457
Epoch 2/30
60792/60792 [==============================] - 76s 1ms/step - loss: 0.3249 - accuracy: 0.9497 - val_loss: 0.3058 - val_accuracy: 0.9588
Epoch 3/30
60792/60792 [==============================] - 77s 1ms/step - loss: 0.3172 - accuracy: 0.9535 - val_loss: 0.3049 - val_accuracy: 0.9585
Epoch 4/30
60792/60792 [==============================] - 77s 1ms/step - loss: 0.3120 - accuracy: 0.9568 - val_loss: 0.3056 - val_accuracy: 0.9592
Epoch 5/30
60792/60792 [==============================] - 77s 1ms/step - loss: 0.3090 - accuracy: 0.9582 - val_loss: 0.3009 - val_accuracy: 0.9617
Epoch 6/30
60792/60792 [==============================] - 77s 1ms/step - loss: 0.3080 - accuracy: 0.9586 - val_loss: 0.3017 - val_accuracy: 0.9616
Epoch 7/30
60792/60792 [==============================] - 77s 1ms/step - loss: 0.3060 - accuracy: 0.9601 - val_loss: 0.3001 - val_accuracy: 0.9622
Epoch 8/30
60792/60792 [==============================] - 77s 1ms/step - loss: 0.3051 - accuracy: 0.9604 - val_loss: 0.3027 - val_accuracy: 0.9615
Epoch 9/30
60792/60792 [==============================] - 77s 1ms/step - loss: 0.3050 - accuracy: 0.9604 - val_loss: 0.3013 - val_accuracy: 0.9616
Epoch 10/30
60792/60792 [==============================] - 77s 1ms/step - loss: 0.3033 - accuracy: 0.9613 - val_loss: 0.2996 - val_accuracy: 0.9629
Epoch 11/30
60792/60792 [==============================] - 78s 1ms/step - loss: 0.3031 - accuracy: 0.9613 - val_loss: 0.3032 - val_accuracy: 0.9603
Epoch 12/30
60792/60792 [==============================] - 77s 1ms/step - loss: 0.3027 - accuracy: 0.9616 - val_loss: 0.3006 - val_accuracy: 0.9624
Epoch 13/30
60792/60792 [==============================] - 78s 1ms/step - loss: 0.3017 - accuracy: 0.9622 - val_loss: 0.2994 - val_accuracy: 0.9626
Epoch 14/30
60792/60792 [==============================] - 78s 1ms/step - loss: 0.3011 - accuracy: 0.9625 - val_loss: 0.3011 - val_accuracy: 0.9616
Epoch 15/30
60792/60792 [==============================] - 78s 1ms/step - loss: 0.3008 - accuracy: 0.9626 - val_loss: 0.2983 - val_accuracy: 0.9635
Epoch 16/30
60792/60792 [==============================] - 78s 1ms/step - loss: 0.3002 - accuracy: 0.9630 - val_loss: 0.2977 - val_accuracy: 0.9639
Epoch 17/30
60792/60792 [==============================] - 78s 1ms/step - loss: 0.3010 - accuracy: 0.9628 - val_loss: 0.2985 - val_accuracy: 0.9638
Epoch 18/30
60792/60792 [==============================] - 78s 1ms/step - loss: 0.2998 - accuracy: 0.9633 - val_loss: 0.2977 - val_accuracy: 0.9640
Epoch 19/30
60792/60792 [==============================] - 78s 1ms/step - loss: 0.2994 - accuracy: 0.9633 - val_loss: 0.2960 - val_accuracy: 0.9645
Epoch 20/30
60792/60792 [==============================] - 78s 1ms/step - loss: 0.2993 - accuracy: 0.9636 - val_loss: 0.2974 - val_accuracy: 0.9638
Epoch 21/30
60792/60792 [==============================] - 78s 1ms/step - loss: 0.3003 - accuracy: 0.9631 - val_loss: 0.2977 - val_accuracy: 0.9639
Epoch 22/30
60792/60792 [==============================] - 78s 1ms/step - loss: 0.2997 - accuracy: 0.9635 - val_loss: 0.2969 - val_accuracy: 0.9640
Epoch 23/30
60792/60792 [==============================] - 78s 1ms/step - loss: 0.3007 - accuracy: 0.9630 - val_loss: 0.2992 - val_accuracy: 0.9628
Epoch 24/30
60792/60792 [==============================] - 78s 1ms/step - loss: 0.2998 - accuracy: 0.9632 - val_loss: 0.3008 - val_accuracy: 0.9626
Epoch 25/30
60792/60792 [==============================] - 78s 1ms/step - loss: 0.2994 - accuracy: 0.9636 - val_loss: 0.2985 - val_accuracy: 0.9635
Epoch 26/30
60792/60792 [==============================] - 78s 1ms/step - loss: 0.2993 - accuracy: 0.9636 - val_loss: 0.2978 - val_accuracy: 0.9633
Epoch 27/30
60792/60792 [==============================] - 78s 1ms/step - loss: 0.2988 - accuracy: 0.9638 - val_loss: 0.2976 - val_accuracy: 0.9637
Epoch 28/30
60792/60792 [==============================] - 78s 1ms/step - loss: 0.2983 - accuracy: 0.9638 - val_loss: 0.2995 - val_accuracy: 0.9633
Epoch 29/30
60792/60792 [==============================] - 1021s 17ms/step - loss: 0.2987 - accuracy: 0.9639 - val_loss: 0.2992 - val_accuracy: 0.9635
Epoch 30/30
60792/60792 [==============================] - 77s 1ms/step - loss: 0.2990 - accuracy: 0.9637 - val_loss: 0.2977 - val_accuracy: 0.9639
-------------------------------------------------------------------------------
----------------- THIS DATA WILL PROBABLY NOT BE USED -------------------------
Final accuracies for the different datasets using validation set number 1 after a total of 30 epoch(s).
* Accuracy on training set: 96.58%
* Accuracy on test set 1: 96.39%
* Accuracy on test set 2: 64.13%
* Accuracy on test set 3: 81.03%
-------------------------------------------------------------------------------
-------------------------------------------------------------------------------
Train on 60792 samples, validate on 54592 samples
Epoch 1/30
60792/60792 [==============================] - 92s 2ms/step - loss: 0.6636 - accuracy: 0.8179 - val_loss: 0.8833 - val_accuracy: 0.7039
Epoch 2/30
60792/60792 [==============================] - 91s 1ms/step - loss: 0.5633 - accuracy: 0.8824 - val_loss: 0.8144 - val_accuracy: 0.7333
Epoch 3/30
60792/60792 [==============================] - 90s 1ms/step - loss: 0.5548 - accuracy: 0.8882 - val_loss: 0.8340 - val_accuracy: 0.7309
Epoch 4/30
60792/60792 [==============================] - 90s 1ms/step - loss: 0.5491 - accuracy: 0.8919 - val_loss: 0.8045 - val_accuracy: 0.7442
Epoch 5/30
60792/60792 [==============================] - 90s 1ms/step - loss: 0.5462 - accuracy: 0.8926 - val_loss: 0.9185 - val_accuracy: 0.6976
Epoch 6/30
60792/60792 [==============================] - 90s 1ms/step - loss: 0.5445 - accuracy: 0.8941 - val_loss: 0.8937 - val_accuracy: 0.7216
Epoch 7/30
60792/60792 [==============================] - 90s 1ms/step - loss: 0.5419 - accuracy: 0.8951 - val_loss: 0.8148 - val_accuracy: 0.7406
Epoch 8/30
60792/60792 [==============================] - 90s 1ms/step - loss: 0.5424 - accuracy: 0.8951 - val_loss: 0.8815 - val_accuracy: 0.7179
Epoch 9/30
60792/60792 [==============================] - 91s 1ms/step - loss: 0.5413 - accuracy: 0.8960 - val_loss: 0.9073 - val_accuracy: 0.7057
Epoch 10/30
60792/60792 [==============================] - 91s 1ms/step - loss: 0.5402 - accuracy: 0.8964 - val_loss: 0.8569 - val_accuracy: 0.7081
Epoch 11/30
60792/60792 [==============================] - 91s 1ms/step - loss: 0.5404 - accuracy: 0.8964 - val_loss: 0.9386 - val_accuracy: 0.7026
Epoch 12/30
60792/60792 [==============================] - 91s 1ms/step - loss: 0.5398 - accuracy: 0.8969 - val_loss: 0.9338 - val_accuracy: 0.7011
Epoch 13/30
60792/60792 [==============================] - 91s 1ms/step - loss: 0.5386 - accuracy: 0.8973 - val_loss: 0.8804 - val_accuracy: 0.7180
Epoch 14/30
60792/60792 [==============================] - 4552s 75ms/step - loss: 0.5398 - accuracy: 0.8967 - val_loss: 0.8787 - val_accuracy: 0.7134
Epoch 15/30
60792/60792 [==============================] - 89s 1ms/step - loss: 0.5389 - accuracy: 0.8972 - val_loss: 0.9015 - val_accuracy: 0.7033
Epoch 16/30
60792/60792 [==============================] - 88s 1ms/step - loss: 0.5386 - accuracy: 0.8973 - val_loss: 0.9438 - val_accuracy: 0.7068
Epoch 17/30
60792/60792 [==============================] - 90s 1ms/step - loss: 0.5378 - accuracy: 0.8977 - val_loss: 0.8121 - val_accuracy: 0.7401
Epoch 18/30
60792/60792 [==============================] - 89s 1ms/step - loss: 0.5391 - accuracy: 0.8973 - val_loss: 0.7806 - val_accuracy: 0.7636
Epoch 19/30
60792/60792 [==============================] - 89s 1ms/step - loss: 0.5383 - accuracy: 0.8976 - val_loss: 0.9030 - val_accuracy: 0.7194
Epoch 20/30
60792/60792 [==============================] - 89s 1ms/step - loss: 0.5374 - accuracy: 0.8978 - val_loss: 0.8644 - val_accuracy: 0.7217
Epoch 21/30
60792/60792 [==============================] - 90s 1ms/step - loss: 0.5380 - accuracy: 0.8977 - val_loss: 0.8347 - val_accuracy: 0.7243
Epoch 22/30
60792/60792 [==============================] - 90s 1ms/step - loss: 0.5379 - accuracy: 0.8977 - val_loss: 0.8830 - val_accuracy: 0.7214
Epoch 23/30
60792/60792 [==============================] - 90s 1ms/step - loss: 0.5381 - accuracy: 0.8977 - val_loss: 1.0105 - val_accuracy: 0.6939
Epoch 24/30
60792/60792 [==============================] - 89s 1ms/step - loss: 0.5376 - accuracy: 0.8977 - val_loss: 0.9599 - val_accuracy: 0.7033
Epoch 25/30
60792/60792 [==============================] - 89s 1ms/step - loss: 0.5381 - accuracy: 0.8977 - val_loss: 0.8518 - val_accuracy: 0.7398
Epoch 26/30
60792/60792 [==============================] - 90s 1ms/step - loss: 0.5382 - accuracy: 0.8975 - val_loss: 0.9031 - val_accuracy: 0.7273
Epoch 27/30
60792/60792 [==============================] - 90s 1ms/step - loss: 0.5381 - accuracy: 0.8977 - val_loss: 0.9445 - val_accuracy: 0.7031
Epoch 28/30
60792/60792 [==============================] - 90s 1ms/step - loss: 0.5377 - accuracy: 0.8977 - val_loss: 0.9105 - val_accuracy: 0.6965
Epoch 29/30
60792/60792 [==============================] - 90s 1ms/step - loss: 0.5369 - accuracy: 0.8982 - val_loss: 0.9030 - val_accuracy: 0.7200
Epoch 30/30
60792/60792 [==============================] - 90s 1ms/step - loss: 0.5381 - accuracy: 0.8978 - val_loss: 0.9719 - val_accuracy: 0.7067
-------------------------------------------------------------------------------
----------------- THIS DATA WILL PROBABLY NOT BE USED -------------------------
Final accuracies for the different datasets using validation set number 2 after a total of 30 epoch(s).
* Accuracy on training set: 90.00%
* Accuracy on test set 1: 89.19%
* Accuracy on test set 2: 70.67%
* Accuracy on test set 3: 79.96%
-------------------------------------------------------------------------------
-------------------------------------------------------------------------------
Train on 60792 samples, validate on 24900 samples
Epoch 1/30
60792/60792 [==============================] - 86s 1ms/step - loss: 0.5435 - accuracy: 0.8574 - val_loss: 0.7158 - val_accuracy: 0.7788
Epoch 2/30
60792/60792 [==============================] - 82s 1ms/step - loss: 0.4430 - accuracy: 0.9165 - val_loss: 0.7546 - val_accuracy: 0.7648
Epoch 3/30
60792/60792 [==============================] - 83s 1ms/step - loss: 0.4348 - accuracy: 0.9208 - val_loss: 0.7417 - val_accuracy: 0.7693
Epoch 4/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4311 - accuracy: 0.9232 - val_loss: 0.7527 - val_accuracy: 0.7696
Epoch 5/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4277 - accuracy: 0.9252 - val_loss: 0.7729 - val_accuracy: 0.7677
Epoch 6/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4256 - accuracy: 0.9263 - val_loss: 0.7677 - val_accuracy: 0.7729
Epoch 7/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4253 - accuracy: 0.9269 - val_loss: 0.7529 - val_accuracy: 0.7698
Epoch 8/30
60792/60792 [==============================] - 82s 1ms/step - loss: 0.4248 - accuracy: 0.9273 - val_loss: 0.7521 - val_accuracy: 0.7702
Epoch 9/30
60792/60792 [==============================] - 80s 1ms/step - loss: 0.4226 - accuracy: 0.9282 - val_loss: 0.7706 - val_accuracy: 0.7676
Epoch 10/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4232 - accuracy: 0.9284 - val_loss: 0.7779 - val_accuracy: 0.7663
Epoch 11/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4227 - accuracy: 0.9282 - val_loss: 0.7851 - val_accuracy: 0.7649
Epoch 12/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4211 - accuracy: 0.9293 - val_loss: 0.7736 - val_accuracy: 0.7712
Epoch 13/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4209 - accuracy: 0.9292 - val_loss: 0.7696 - val_accuracy: 0.7717
Epoch 14/30
60792/60792 [==============================] - 83s 1ms/step - loss: 0.4211 - accuracy: 0.9295 - val_loss: 0.7815 - val_accuracy: 0.7727
Epoch 15/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4195 - accuracy: 0.9302 - val_loss: 0.7680 - val_accuracy: 0.7742
Epoch 16/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4193 - accuracy: 0.9302 - val_loss: 0.7637 - val_accuracy: 0.7711
Epoch 17/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4191 - accuracy: 0.9303 - val_loss: 0.7863 - val_accuracy: 0.7673
Epoch 18/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4187 - accuracy: 0.9305 - val_loss: 0.7871 - val_accuracy: 0.7686
Epoch 19/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4187 - accuracy: 0.9305 - val_loss: 0.7730 - val_accuracy: 0.7802
Epoch 20/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4191 - accuracy: 0.9304 - val_loss: 0.7953 - val_accuracy: 0.7735
Epoch 21/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4194 - accuracy: 0.9304 - val_loss: 0.7711 - val_accuracy: 0.7762
Epoch 22/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4185 - accuracy: 0.9307 - val_loss: 0.7800 - val_accuracy: 0.7698
Epoch 23/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4183 - accuracy: 0.9308 - val_loss: 0.7747 - val_accuracy: 0.7706
Epoch 24/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4180 - accuracy: 0.9310 - val_loss: 0.8009 - val_accuracy: 0.7711
Epoch 25/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4193 - accuracy: 0.9305 - val_loss: 0.7877 - val_accuracy: 0.7783
Epoch 26/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4181 - accuracy: 0.9310 - val_loss: 0.8029 - val_accuracy: 0.7687
Epoch 27/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4185 - accuracy: 0.9308 - val_loss: 0.7975 - val_accuracy: 0.7675
Epoch 28/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4175 - accuracy: 0.9313 - val_loss: 0.7793 - val_accuracy: 0.7735
Epoch 29/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4179 - accuracy: 0.9311 - val_loss: 0.7798 - val_accuracy: 0.7735
Epoch 30/30
60792/60792 [==============================] - 81s 1ms/step - loss: 0.4182 - accuracy: 0.9309 - val_loss: 0.7894 - val_accuracy: 0.7724
-------------------------------------------------------------------------------
----------------- THIS DATA WILL PROBABLY NOT BE USED -------------------------
Final accuracies for the different datasets using validation set number 3 after a total of 30 epoch(s).
* Accuracy on training set: 93.29%
* Accuracy on test set 1: 92.74%
* Accuracy on test set 2: 61.09%
* Accuracy on test set 3: 77.24%
-------------------------------------------------------------------------------
-------------------------------------------------------------------------------
</p>

